src/bin/ipa_bench/models.rs

```diff
@@ -109,57 +111,169 @@ impl SecretSharable for u64 {
-        )
+impl AddAssign for EventTimestamp {
+    // #[allow(clippy::assign_op_pattern)]
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs;
+    }
+}
+
+impl PartialOrd for EventTimestamp {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(u64::from(*self).cmp(&u64::from(*other)))
+    }
+}
+
+impl From<EventTimestamp> for u64 {
+    fn from(v: EventTimestamp) -> Self {
+        v.0 % (EventTimestamp::MAX + 1)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Event {
+    // An identifier, set in the user agent, which identifies an individual person. This must never be released (beyond
+    /// the match key provider) to any party in unencrypted form. For the purpose of this tool, however, the value is in
+    /// clear and not secret shared.
+    pub matchkey: MatchKey,
+
+    /// An identifier, specified by the report collector, to denote if a given pair of source and trigger events can be
+    /// attributed (beyond having the same match key.) If None, a trigger event will be attributed to all source events.
+    pub attribution_constraint_id: Option<Number>,
+
+    /// A timestamp in which this event is generated.
+    pub timestamp: EventTimestamp,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum GenericReport {
+    /// An event produced on websites/apps when a user interacts with an ad (i.e. impression, click).
+    Source {
+        event: Event,
+
+        /// A key, specified by the report collector, which allows for producing aggregates across many groups (or breakdowns.)
+        /// This field is only valid if is_trigger_report is [false].
+        breakdown_key: Number,
+    },
+
+    /// An event produced on websites/apps when a user takes an action (i.e. product view, purchase).
+    Trigger {
+        event: Event,
+
+        /// The value of the trigger report to be aggregated. This field is only valid if is_trigger_report is [true].
+        value: Number,
+    },
+}
+
+/// Serialize trigger and source reports to the same format. This prevents information leakage by ensuring the helper
+/// parties are unable to differentiate between source and trigger reports throughout the entire protocol.
+impl Serialize for GenericReport {
+    // TODO: In production, fields in GenericReport will be encrypted and secret shared.
+
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            GenericReport::Source {
+                event,
+                breakdown_key,
+            } => {
+                let mut state = serializer.serialize_struct("GenericReport", 7)?;
+                state.serialize_field("matchkey", &event.matchkey)?;
+                state.serialize_field(
+                    "attribution_constraint_id",
+                    &event.attribution_constraint_id,
+                )?;
+                state.serialize_field("epoch", &event.timestamp.epoch())?;
```

Comment:
can we implement `Serialize` for `EpochTimestamp` and use it here? at the very least we can fit both epoch and offset in a single `u32` or `[u8; 3]`

Reply:
Yes, let me work on improving `Serialize` and `Deserialize` implementation in the next diff.

I think we need multiple `EventTimestamp` (or `GenericReport` for that matter) variants. One that is used in MPC, which should implement `Serialize` that mixes epoch and offset to `[u8; 3]', and another for outside MPC such as synthetic data generation (e2e test input). For the latter, I would like to keep the serialization simple.

Reply:
That was another reason that led me to use trait so that we can have two different EventTimestamp structs. Let me implement `Serialize` based on your suggestion first and see if that works for all use cases including an E2E test. If not, we can make modifications then.

Reply:
how about taking the serialization out of this diff? I think apart from it, it is in a good shape.